http/basic/src/main/java/io/cloudevents/http/HttpMessageFactory.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudevents.http;
+
+import io.cloudevents.CloudEvent;
+import io.cloudevents.core.format.EventFormat;
+import io.cloudevents.core.message.MessageReader;
+import io.cloudevents.core.message.MessageWriter;
+import io.cloudevents.core.message.impl.GenericStructuredMessageReader;
+import io.cloudevents.core.message.impl.MessageUtils;
+import io.cloudevents.core.message.impl.UnknownEncodingMessageReader;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+public class HttpMessageFactory {
+
+    public static MessageReader createReader(Map<String,String> headers, byte[] body) throws IllegalArgumentException {
+        return MessageUtils.parseStructuredOrBinaryMessage(
+            () -> headers.get("Content-Type"),
```

Comment:
I guess this line and the one two below are where case-sensitivity comes into play? You could write a small helper method:

```
  private static <V> V getIgnoringCase(Map<String, V> map, String key) {
    for (Map.Entry<String, V> entry : map.entrySet()) {
      if (key.equalsIgnoreCase(entry.getKey())) {
        return entry.getValue();
      }
    }
  }
```

Or, you could copy the input map into a `new TreeMap<>(String.CASE_INSENSITIVE_ORDER)`.

Reply:
Indeed it looks like this is the only place where capitalization matters. I don't like idea of creating `new` cases insensitive map just to do two `gets()`. That linear search might be better. I could do following:
```java
    public static MessageReader createReader(Map<String,String> headers, byte[] body) throws IllegalArgumentException {
        String _ct = null;
        String _sv = null;

        for (Map.Entry<String,String> kvp: headers.entrySet()) {
            if (kvp.getKey().equalsIgnoreCase(CONTENT_TYPE)) {
                _ct = kvp.getValue();
            } else if (kvp.getKey().equalsIgnoreCase(CloudEventsHeaders.SPEC_VERSION)) {
                _sv = kvp.getValue();
            }
        }

        final String specversion = _sv;
        final String contentType = _ct;

        return MessageUtils.parseStructuredOrBinaryMessage(
            () -> contentType,
            format -> new GenericStructuredMessageReader(format, body),
            () -> specversion,
            sv -> new HttpMessageReader(sv, headers, body),
            UnknownEncodingMessageReader::new
        );
    }
```

Reply:
Other option would be delegate that to caller. Javadoc could state that the map has to be case insensitive, but I don't think that is good idea as users might miss that. wdyt @slinkydeveloper

Reply:
> Other option would be delegate that to caller. Javadoc could state that the map has to be case insensitive, but I don't think that is good idea as users might miss that. wdyt @slinkydeveloper

I think we should provide a case insensitive solution by ourselves, at least at the beginning. This module should be a best-effort but correct solution, not an optimized one: if you want the optimized solution, implement the reader/writer for your specific data structure, that already solves these problems by itself.

Reply:
If I am going to use the linear search I could as well not use `java.util.Map`, but `Consumer<BiConsumer<String,String>>`
```java
public static MessageReader createReader(Consumer<BiConsumer<String,String>> forEachHeader, byte[] body) throws IllegalArgumentException {
        final String[] _ct = {null};
        final String[] _sv = {null};

        forEachHeader.accept((k, v) -> {
            if (k.equalsIgnoreCase(CONTENT_TYPE)) {
                _ct[0] = v;
            } else if (k.equalsIgnoreCase(CloudEventsHeaders.SPEC_VERSION)) {
                _sv[0] = v;
            }
        });

        final String specversion = _sv[0];
        final String contentType = _ct[0];

        return MessageUtils.parseStructuredOrBinaryMessage(
            () -> contentType,
            format -> new GenericStructuredMessageReader(format, body),
            () -> specversion,
            sv -> new HttpMessageReader(sv, forEachHeader, body),
            UnknownEncodingMessageReader::new
        );
    }
```

Reply:
What's more desirable?
```java
public static MessageReader createReader(Map<String,String> headers, byte[] body) throws IllegalArgumentException;
// VS
public static MessageReader createReader(Consumer<BiConsumer<String,String>> forEachHeader, byte[] body) throws IllegalArgumentException;

Reply:
I think that biconsumer might suit well for multimaps too

Reply:
And we can have the `Map` signature as well

Reply:
The advantage of the `Consumer<BiConsumer<String, String>>` solution is that it sidesteps the question of whether to use `Map<String, String>` or `Map<String, List<String>>`, but the disadvantage is that it's kind of daunting for users. If I see a parameter `Consumer<BiConsumer<String, String>>` my first reaction is to panic. :-)
I think it would be OK, if the javadoc showed how to use it with the two `Map` cases, something like this:

```
Map<String, String> headers = ...;
MessageReader reader = HttpMessageFactory.createReader(
    forEach -> headers.forEach(forEach), body);
```
```
Map<String, List<String>> headers = ...;
Consumer<BiConsumer<String, String>> consumer = forEach ->
    headers.forEach((key, values) -> values.forEach(value -> forEach.accept(key, value)));
MessageReader reader = HttpMessageFactory.createReader(forEach, body);
```

As @slinkydeveloper suggests, we could also have both `Consumer` and `Map` overloads. The `Map` overload could have javadoc that says it's equivalent to one of the code extracts above (depending on what flavour of `Map` it is). Then the `Consumer` overload javadoc could show the other code extract.